### ISSUE TITLE
What is the best (free or paid) hosting for Ogar

### ISSUE THREAD

**Issue by lukerayman** [OgarProject/Ogar#581]
Currently I'm testing Cloud9. They're good because they're free and easy to configure. When there is less than 10 players then it works fine. But when I see about 16 people on the map, the game is slowing down. It's getting unplayable. And I wanted to make place for 256 people... It looks impossible for Cloud9. 

Do you know what is the best option for Ogar. May be paid, but not too expensive. I discovered that Agar.io servers works on OVH.

**Comment by F0RIS** [comment]
I see you tested and got bad result on Cloud9 too) I don't know what is the best hosting. But as for me, currently i am using DigitalOcean's cheapest VPS. It can hold around 90 players. But there is memory leak issue

**Comment by Andrews54757** [comment]
@F0RIS use a garbage collection system for memory leak

**Comment by lukerayman** [comment]
Thanks @F0RIS I will test DigitalOcean today.

**Comment by lukerayman** [comment]
@F0RIS You're right, there is a memory leak problem. My servers is getting killed like every 30 minutes. I tried with Garbage Collection System - didn't help. :(

**Comment by typngeek** [comment]
I have successfully hosted a server with 120 concurrent players on experimental. But once you hit 80 players you start getting some serious lag, which is no fun for players. Getting any further beyond this without lag is something I haven't figured out yet. 

When you get to that level, you hit a big CPU bottleneck because node.js is single-threaded, which is why I'm eagerly awaiting further development of the java version. I host with SSDNodes (https://lowendbox.com/blog/ssdnodes-ssd-openvz-vps-starting-at-2-99month-dallas-tx/) on the $7 a month package because it has enough RAM (4GB) and they run Intel E5v3 processors, which are pretty darn powerful for single-threaded workloads. Not affiliated with the company, but love them to pieces.

To overcome any memory leak issues, I set up a crontab job that runs at 3am every morning to attach to the Ogar screen, kill the process, and restart it. Even with this though, I frequently find myself getting notified that the Agar.io server is dead a few times a day. It was such a headache that I ended up eventually passing along the responsibility to someone else.

Let me know if you guys have any questions.

![image](https://cloud.githubusercontent.com/assets/16531060/15449999/a601979e-1f5c-11e6-8b24-6bae4613a9bd.png)

**Comment by lukerayman** [comment]
@typngeek I have a question. I opened this site: https://www.ssdnodes.com/cloud/startup/ And I can't find any cheaper offer than $9.99. How did you get this $7 a month price?

**Comment by EnTerr** [comment]
Wat?
Memory leak - how is this possible in a memory-managed language like JavaScript?!
Is this an issue of the JS runtime or Ogar somehow keeps hogging references to objects that should be dead?

**Comment by typngeek** [comment]
@lukerayman, @EnTerr is right about the link. Take a look at it. 

As for the memory leak stuff, I never encountered problems once I upgraded to 4GB of RAM, but when I was originally hosting with a 2GB OVH VPS, I frequently had crashes because of the memory limitations. I could probably host like 20 or 30 players before it started lagging and eventually crashed. Whenever it crashed, I had to SSH back into the VPS and restart it, which was a royal pain especially when I wasn't near a computer.

**Comment by EnTerr** [comment]
So not necessarily memory **leak**? Sounds more like Ogar or Node.js is a memory **hog** :)

Can't one put a watchdog to restart the server when it hangs? Not much relief for those whose game got killed but still, no need of manual ssh. 

Actually how does it "crash"? - does the V8 engine hang or core dump or what

**Comment by lukerayman** [comment]
OK, the link works now. But does it make sense to buy VSP with multi-core processor for Node.js? As far as I know Node.js runs in one thread only.

**Comment by typngeek** [comment]
Perhaps. Never actually got down to the bottom of it.

I tried using forever (It's on github) to restart the nodejs process when it crashes but the problem I noticed was that I no longer could enter commands in the console, which is a deal breaker. The auto restart works though, so if there is a work around for the commands...

On Sun, May 22, 2016 at 7:09 PM, EnTerr < notifications@github.com [notifications@github.com] > wrote:
So not necessarily memory leak ? Sounds more like Ogar or Node.js is a memory hog :)

Can't one put a watchdog to restart the server when it hangs? Not much relief for those whose game got killed but still, no need of manual ssh.

Actually how does it "crash"? - does the V8 engine hang or core dump or what

**Comment by F0RIS** [comment]
@typngeek if somebody interesting  - i wrote shell script which work like daemon. It reastarting server if it fall for some reason. Also there is some simple manager for dealing with serveral Ogar servers. I would be very happy for any kind of contributing and if those files would be usefull for someone
https://github.com/F0RIS/Ogar-servers-manager

**Comment by F0RIS** [comment]
@typngeek it seems that few days ago i saw on github project, which allow to enter console commands remotely. Find it and try)

**Comment by jollesnaas** [comment]
You can rent a VPS (Windows or Linux) of me: 4 euro/month/512MB RAM or 8 euro/month/1GB RAM, 20 euro/Year/512MB RAM or 40 euro/year/1GB RAM.

If you want to know more or order, please contact me on skype: jolle.snaas
*Free test included!

**Comment by Andrews54757** [comment]
i dont trust this . dont fall for it others! Only give money to trusted people

**Comment by Andrews54757** [comment]
ovh is cheaper
